arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -0,0 +1,192 @@
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Valery Mironov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Metrics/ClusterMetricsFeature.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/debugging.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/ClusterMethods.h"
+#include "Cluster/ServerState.h"
+#include "Containers/FlatHashSet.h"
+#include "Logger/LoggerFeature.h"
+#include "Metrics/Metric.h"
+#include "Metrics/MetricsFeature.h"
+#include "Network/NetworkFeature.h"
+#include "ProgramOptions/ProgramOptions.h"
+#include "ProgramOptions/Section.h"
+#include "RestServer/QueryRegistryFeature.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+#include "Scheduler/SchedulerFeature.h"
+#include "Statistics/StatisticsFeature.h"
+#include "StorageEngine/EngineSelectorFeature.h"
+
+namespace arangodb::metrics {
+using namespace std::chrono_literals;
+
+ClusterMetricsFeature::ClusterMetricsFeature(Server& server)
+    : ArangodFeature{server, *this} {
+  setOptional();
+  startsAfter<ClusterFeature>();
+  startsAfter<NetworkFeature>();
+  startsAfter<SchedulerFeature>();
+}
+
+void ClusterMetricsFeature::collectOptions(
+    std::shared_ptr<options::ProgramOptions> options) {
+  options
+      ->addOption(
+          "--server.cluster-metrics-timeout",
+          "Seconds that will be between cross cluster calls that collect "
+          "metrics",
+          new options::UInt32Parameter(&_timeout),
+          arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+      .setIntroducedIn(310000);
+}
+
+void ClusterMetricsFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> /*options*/) {
+  if (!ServerState::instance()->isCoordinator()) {
+    disable();
+  }
+}
+void ClusterMetricsFeature::beginShutdown() {
+  _count.store(std::numeric_limits<int32_t>::min() + 1,
+               std::memory_order_relaxed);
+}
+
+void ClusterMetricsFeature::asyncUpdate() {
+  if (isEnabled() && _count.fetch_add(1, std::memory_order_acq_rel) == 0) {
+    scheduleUpdate();
+  }
+}
+
+void ClusterMetricsFeature::scheduleUpdate() noexcept {
+  auto now = std::chrono::steady_clock::now();
+  auto next = _lastUpdate + std::chrono::seconds{_timeout};
+  _handle = SchedulerFeature::SCHEDULER->queueDelayed(
+      RequestLane::CLUSTER_INTERNAL, (now < next ? next - now : 0ns),
+      [this](bool canceled) {
+        if (canceled) {
+          return;
+        }
+        _count.store(1, std::memory_order_relaxed);
```

Comment:
Does this store have to be synced with the fetch_add() in line 79?

Reply:
Not sure. But looks like this line could cancel the effect of `beginShutdown` if timing is unlucky.